### PR TITLE
Update exception handling support to current proposal

### PR DESCRIPTION
**Pull request by takikawa**
This PR updates the support of exception handling to the latest proposal (that is compatible with future 2-phase exception handling) described in https://github.com/WebAssembly/exception-handling/pull/137 and https://github.com/WebAssembly/exception-handling/pull/143.

* Adds back tagged `catch $e`, `catch_all`, and `rethrow N` from a previous version of wabt, but with updates to match the current spec (e.g., `catch_all` shares an opcode with `else`, `rethrow`'s depth indexes only catch blocks, etc).
* Adds `unwind` and `delegate` instructions.
* Removes `exnref` and `br_on_exn`.
* Updates relevant tests.

There are some details that could still change (e.g., maybe how `delegate`'s depth is validated), but I'd be happy to submit further PRs if the spec details change.

### CODE REVIEW

File: test/dump/try-delegate.txt [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+try-delegate.wasm:	file format wasm 0x1
+
+Code Disassembly:
+
+00001f func[0]:
+ 000020: 06 7f                      | try i32
+ 000022: 06 40                      |   try
+ 000024: 01                         |     nop
+ 000025: 18 01                      |     delegate 1
```
Comment (binji) [inline test/dump/try-delegate.txt:91]: looks like delegate should dedent here?

File: test/dump/try-unwind.txt [hunk 2]
```diff
@@ -55,41 +54,37 @@
...
- 000028: 1a                         |     drop
- 000029: 0b                         |   end
- 00002a: 41 00                      |   i32.const 0
- 00002c: 0b                         | end
- 00002d: 1a                         | drop
- 00002e: 0b                         | end
+ 000020: 06 7f                      | try i32
+ 000022: 01                         |   nop
+ 000023: 41 07                      |   i32.const 7
+ 000025: 0a                         |   unwind
```
Comment (binji) [inline test/dump/try-unwind.txt:84]: unwind should dedent too, then reindent like `catch` or `else`

**Review by binji — APPROVED** [review]
lgtm, thanks for doing this! Sorry for the long delay on review

File: test/dump/try-delegate.txt (hunk 1 above)
Comment (takikawa) [reply test/dump/try-delegate.txt]: Thanks for catching this (& `unwind` too)! Latest commit should fix this issue.

**Comment by dschuff** [comment]
I just noticed (after the discussion in #1604, thanks for pointing it out) that this PR doesn't actually solve the problem that made me notice the opcode problem, (because I was only using the disassembler). Namely `wasm-objdump` prints opcode `0x5` as `else` rather than `catch-all` even in the context of try. Presumably it's just going directly from `opcode.def`, (which doesn't have `catch_all` in this PR). But maybe let's not fix this until we decide whether we want to keep the opcodes as they are or change them.

**Comment by aheejin** [comment]
By the way can I merge this? If the opcode is the issue we can fix it later.

**Comment by takikawa** [comment]
@aheejin Sure, no issues on my end with having it merged. AFAIK I've addressed all the comments in review. Thanks!

I'd also be happy to submit a PR afterwards if the opcode changes.

**Pull request merged**